packages/hardhat/contracts/CGTSwap.sol

```diff
@@ -0,0 +1,158 @@
+        _requireNotPaused();
+        _pause();
+    }
+
+    /// @notice Unpause token swaps
+    function unpause() external onlyOwner {
+        _requirePaused();
+        _unpause();
+    }
+
+    /**
+     * @notice Swap CGT for stablecoin
+     * @param amountIn input amount of CGT
+     */
+    function swap(uint256 amountIn) 
+        external 
+        whenNotPaused
+        nonReentrant
+    {
+        require(amountIn > 0, "Invalid amount");
+        _swap(amountIn);
+    }
+
+    /// @notice Transfers fee collected from the contract to the owner
+    function withdrawFeeCollected() external onlyOwner {
+        TransferHelper.safeTransfer(
+            address(stable), 
+            owner(),
+            totalFeeCollected
+        );
+        totalFeeCollected = 0;
+    }
+
+    /**
+     * @notice Transfers tokens from the contract to the owner
+     * @param token address of token to withdraw
+     * @param amount amount of tokens to withdraw
+     */
+    function withdrawTokens(address token, uint256 amount) external onlyOwner {
+        require(token != address(0), "Invalid token address");
+        TransferHelper.safeTransfer(
+            token, 
+            owner(),
+            amount
+        );
+    }
+
+    /**
+     * @notice Calculate estimated amount of stablecoin received
+     * @dev Uses Chainlink XAU price feed
+     * @param amountIn input amount of CGT
+     * @return estimated stablecoin amount out
+     */
+    function quoteStablecoinAmount(uint256 amountIn) public view returns (uint256) {
+        uint256 stableAmount = (amountIn * uint256(getLatestXAU_USDPrice()) * 3215075) / (10**8 * 10**8);
+        uint256 adjustedAmount = (stableAmount / DECIMAL_FACTOR) + round(stableAmount, DECIMAL_FACTOR);
+        uint256 netStablecoinAmount = (adjustedAmount * (10000 - margin)) / 10000;
+        return netStablecoinAmount;
+    }
+
+    /**
+     * @notice View stablecoin balance of the contract
+     * @return stablecoin balance
+     */
+    function getStablecoinBalance() public view returns (uint256) {
+        return stable.balanceOf(address(this));
+    }
+
+    /**
+     * @notice Returns the latest XAU-USD price
+     * @dev Calls XAU-USD data feed
+     * @return XAU-USD price
+     */
+    function getLatestXAU_USDPrice() public view returns (int256) {
+        (
+            /*uint80 roundID*/,
+            int price,
+            /*uint startedAt*/,
+            /*uint timeStamp*/,
+            /*uint80 answeredInRound*/
+        ) = priceFeed.latestRoundData();
+        return price;
+    }
+
+    function _swap(uint256 amountIn) private {
+        TransferHelper.safeTransferFrom(address(CGT), msg.sender, address(this), amountIn);
+        uint256 amountOut = quoteStablecoinAmount(amountIn);
+        totalFeeCollected += (amountOut *  margin) / (10000 - margin);
+
+        require(amountOut <= stable.balanceOf(address(this)), "Insufficient stablecoin available");
```

Comment:
this would fail in the transfer step that is next? So it is redundant.

Reply:
Yep that's right. But just did so in order to improve readability of the revert string. Let me know if i should remove it.

Reply:
I think lets remove it, since this is mainnet and additional gas costs are expensive

Reply:
Okay, done.